src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java

```diff
@@ -0,0 +1,114 @@
+    this.binaryValuesStreamer = binaryValuesStreamer;
+    this.idIndexIteratorConverter = idIndexIteratorConverter;
+  }
+
+  /**
+   * Creates a data-flow tree from the pruned entity tree. The constructed tree will be rooted with the outputEntity
+   * passed in.
+   * @param prunedEntityTree Entity tree pruned to only active nodes based on output and filters.
+   * @param outputEntity Entity for which variables are output.
+   * @param filters Filters to apply to entity stream.
+   * @param outputVariables Variables to return for each output entity.
+   * @param study Study for which
+   * @return
+   */
+  public DataFlowMapReduceTree create(TreeNode<Entity> prunedEntityTree,
+                                      Entity outputEntity,
+                                      List<Filter> filters,
+                                      List<Variable> outputVariables,
+                                      Study study) {
+    if (!outputVariables.stream().allMatch(var -> var.getEntity().equals(outputEntity))) {
+      throw new IllegalArgumentException("All output variables must be associated with output entity.");
+    }
+    try {
+      LOG.info("Creating map-reduce data-flow tree rooted with output entity: " + outputEntity.getId());
+      final TreeNode<Entity> parent = prunedEntityTree.findFirst(entity -> entity.equals(outputEntity));
+      final SubsettingJoinNode root = rerootTree(prunedEntityTree, parent, null, filters, study);
+      final List<Iterator<VariableValueIdPair<?>>> outputStreams = outputVariables.stream()
+          .map(Functions.fSwallow(varWithValues ->
+              // Any variables without values (i.e. categories) should fail validation upstream.
+              binaryValuesStreamer.streamIdValuePairs(study, (VariableWithValues<?>) varWithValues)))
+          .collect(Collectors.toList());
+      final Iterator<VariableValueIdPair<List<Long>>> ancestorStreams = outputEntity.getAncestorEntities().isEmpty()
+          ? null
+          : binaryValuesStreamer.streamAncestorIds(outputEntity, study);
+      return new DataFlowMapReduceTree(outputStreams, root, ancestorStreams);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to open files to stream binary data", e);
+    }
+  }
+
+  /**
+   * Recursively depth-first traverse each child and then the parent.
+   *
+   * @param originalRoot The original entity tree's root. This is needed to find the parent of our current node.
+   * @param currentTraversalNode Node that is currently being traversed.
+   * @param previousNode Node that was previously traversed. We keep track of this to avoid traversing from whence we came.
+   * @param filters List of all filters in request, used in construction of the {@link SubsettingJoinNode} corresponding to this node.
+   * @param study Study that subsetting data is requested for.
+   * @return
+   */
+  private SubsettingJoinNode rerootTree(TreeNode<Entity> originalRoot,
```

Comment:
i am having trouble understanding this method too.

Reply:
Some direction might be helpful here or perhaps a conversation